File: leetcode/python/majorityElement.py


# leetcode.com majorityElement

class Solution:
    def majorityElement(self, num):
        d = {}
        size = len(num)

        for i in num:
            if i in d:
                d[i] += 1
            else:
                d[i] = 1

            if d[i] > size / 2:
                return i

File: leetcode/java/src/main/java/com/vkeonline/leetcode/may/MajorityElement.java
package com.vkeonline.leetcode.may;

public class MajorityElement {
    public int majorityElement(int[] nums) {
        return findCandidate(nums);
    }

    boolean isMajority(int nums[]) {
        int candidate = findCandidate((nums));
        int i, count = 0;
        for (i = 0; i < nums.length; i++) {
            if (nums[i] == candidate)
                count++;
        }
        if (count > nums.length / 2)
            return true;
        else
            return false;
    }

    private int findCandidate(int nums[]) {
        int majorIndex = 0, count = 1;
        int i;
        for (i = 1; i < nums.length; i++) {
            if (nums[majorIndex] == nums[i])
                count++;
            else
                count--;
            if (count == 0) {
                majorIndex = i;
                count = 1;
            }
        }
        return nums[majorIndex];
    }
}
